planner/core/rule_derive_topn_from_window.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/planner/util"
+)
+
+// deriveTopNFromWindow pushes down the topN or limit. In the future we will remove the limit from `requiredProperty` in CBO phase.
+type deriveTopNFromWindow struct {
+}
+
+func appendDerivedTopNTrace(topN LogicalPlan, opt *logicalOptimizeOp) {
+	child := topN.Children()[0]
+	action := func() string {
+		return fmt.Sprintf("%v_%v top N added below  %v_%v ", topN.TP(), topN.ID(), child.TP(), child.ID())
+	}
+	reason := func() string {
+		return fmt.Sprintf("%v filter on row number", topN.TP())
+	}
+	opt.appendStepToCurrent(topN.ID(), topN.TP(), reason, action)
+}
+
+/*
+		Check the following pattern of filter over row number window function:
+	  - Filter is simple condition of row_number < value or row_number <= value
+	  - The window function is a simple row number
+	  - With default frame: rows between current row and current row. Check is not necessary since
+	    current row is only frame applicable to row number
+	  - No partition
+	  - Child is a data source.
```

Comment:
Customer's case there is filter on top of data source, but we can do it in nest PR.